File: internal/watcher/watcher.go
package watcher

import (
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/fsnotify/fsnotify"
)

// Watcher observes a workspace and calls nudge once the user has edited
// files and then gone idle. It is a hint source only: file events never
// trigger reviews directly, /done stays authoritative, and any watch
// failure degrades to silence.
type Watcher struct {
	fs    *fsnotify.Watcher
	root  string
	idle  time.Duration
	nudge func()
}

var ignoredDirs = map[string]bool{
	".git":         true,
	".nina":        true,
	"node_modules": true,
	"__pycache__":  true,
	".venv":        true,
	"venv":         true,
	".idea":        true,
	".vscode":      true,
}

func Start(dir string, idle time.Duration, nudge func()) (*Watcher, error) {
	fs, err := fsnotify.NewWatcher()
	if err != nil {
		return nil, err
	}
	w := &Watcher{fs: fs, root: dir, idle: idle, nudge: nudge}
	if err := w.addRecursive(dir); err != nil {
		fs.Close()
		return nil, err
	}
	go w.loop()
	return w, nil
}

func (w *Watcher) Close() error {
	return w.fs.Close()
}

func (w *Watcher) loop() {
	var timer *time.Timer
	var timerC <-chan time.Time
	for {
		select {
		case event, ok := <-w.fs.Events:
			if !ok {
				return
			}
			if w.ignored(event.Name) {
				continue
			}
			if event.Has(fsnotify.Create) {
				if info, err := os.Stat(event.Name); err == nil && info.IsDir() {
					_ = w.addRecursive(event.Name)
					continue
				}
			}
			if timer == nil {
				timer = time.NewTimer(w.idle)
				timerC = timer.C
			} else {
				if !timer.Stop() {
					select {
					case <-timerC:
					default:
					}
				}
				timer.Reset(w.idle)
			}
		case <-timerC:
			timer = nil
			timerC = nil
			w.nudge()
		case _, ok := <-w.fs.Errors:
			if !ok {
				return
			}
		}
	}
}

func (w *Watcher) addRecursive(dir string) error {
	return filepath.WalkDir(dir, func(path string, entry os.DirEntry, err error) error {
		if err != nil {
			return nil
		}
		if !entry.IsDir() {
			return nil
		}
		if path != w.root && ignoredDirs[entry.Name()] {
			return filepath.SkipDir
		}
		return w.fs.Add(path)
	})
}

func (w *Watcher) ignored(path string) bool {
	rel, err := filepath.Rel(w.root, path)
	if err != nil || strings.HasPrefix(rel, "..") {
		return true
	}
	for part := range strings.SplitSeq(filepath.ToSlash(rel), "/") {
		if ignoredDirs[part] {
			return true
		}
	}
	base := filepath.Base(path)
	// Editor noise: vim/emacs swap, temp, and lock files.
	return strings.HasSuffix(base, "~") ||
		strings.HasSuffix(base, ".swp") ||
		strings.HasSuffix(base, ".swo") ||
		strings.HasSuffix(base, ".tmp") ||
		strings.HasPrefix(base, ".#") ||
		strings.HasPrefix(base, "#")
}

File: python/src/nina/watcher.py
from __future__ import annotations

import threading
from collections.abc import Callable
from pathlib import Path

from watchdog.events import FileSystemEvent, FileSystemEventHandler
from watchdog.observers import Observer

IGNORED_DIRS = {
    ".git",
    ".nina",
    "node_modules",
    "__pycache__",
    ".venv",
    "venv",
    ".idea",
    ".vscode",
}


class Watcher:
    def __init__(self, dir: str, idle_seconds: float, nudge: Callable[[], None]) -> None:
        self._root = Path(dir).resolve()
        self._idle_seconds = idle_seconds
        self._nudge = nudge
        self._lock = threading.Lock()
        self._timer: threading.Timer | None = None
        self._observer = Observer()
        self._observer.schedule(
            _Handler(self._root, self._on_activity), str(self._root), recursive=True
        )
        self._observer.start()

    def close(self) -> None:
        self._observer.stop()
        self._observer.join(timeout=2)
        with self._lock:
            if self._timer is not None:
                self._timer.cancel()
                self._timer = None

    def _on_activity(self) -> None:
        with self._lock:
            if self._timer is not None:
                self._timer.cancel()
            self._timer = threading.Timer(self._idle_seconds, self._fire)
            self._timer.daemon = True
            self._timer.start()

    def _fire(self) -> None:
        with self._lock:
            self._timer = None
        self._nudge()


class _Handler(FileSystemEventHandler):
    def __init__(self, root: Path, on_activity: Callable[[], None]) -> None:
        self._root = root
        self._on_activity = on_activity

    def on_any_event(self, event: FileSystemEvent) -> None:
        if event.is_directory:
            return
        path = Path(str(event.src_path))
        if self._ignored(path):
            return
        self._on_activity()

    def _ignored(self, path: Path) -> bool:
        try:
            rel = path.resolve().relative_to(self._root)
        except ValueError:
            return True
        if any(part in IGNORED_DIRS for part in rel.parts):
            return True
        name = path.name
        return (
            name.endswith(("~", ".swp", ".swo", ".tmp"))
            or name.startswith(".#")
            or name.startswith("#")
        )
